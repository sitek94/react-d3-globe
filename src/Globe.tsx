import * as React from 'react';
import { geoOrthographic, geoPath, select } from 'd3';

import { Features, getFeatures } from './getFeatures';

export interface GlobeProps {
  oceanColor?: string;
  landColor?: string;

  /**
   * A shorthand for height and width of the SVG element
   */
  size?: number;

  /**
   * Height of the SVG
   */
  height?: number;

  /**
   * Width of the SVG
   */
  width?: number;

  /**
   *  Scale factor to be used for the projection
   */
  scale?: number;

  /**
   * A point specified as a two-dimensional array [longitude, latitude] in degrees.
   * This will be the projectionâ€™s center.
   */
  center?: [number, number];

  /**
   * The x-axis rotation angle in degrees
   */
  rotateX?: number;

  /**
   * The y-axis rotation angle in degrees
   */
  rotateY?: number;

  /**
   * The z-axis rotation angle in degrees
   */
  rotateZ?: number;
}

export function Globe({ size = 400, ...rest }: GlobeProps) {
  const {
    oceanColor = '#eaedee',
    landColor = '#17181d',
    height = size,
    width = size,
    scale = size / 2,
    center = [0, 0],
    rotateX = 0,
    rotateY = 0,
    rotateZ = 0,
  } = rest;

  // Variables

  const centerX = width / 2;
  const centerY = height / 2;
  const circleR = scale;
  const rotation: [number, number, number] = [rotateX, rotateY, rotateZ];

  // State

  const divRef = React.useRef<HTMLDivElement>(null);
  const [features, setFeatures] = React.useState<Features>([]);

  React.useEffect(() => {
    getFeatures().then(setFeatures).catch(console.error);
  }, []);

  // Projection
  const projection = React.useMemo(
    () =>
      geoOrthographic()
        .scale(scale)
        .center(center)
        .rotate(rotation)
        .translate([centerX, centerY]),
    [scale, center, rotation, centerX, centerY]
  );

  // Path generator
  const pathGenerator = geoPath().projection(projection);

  // Update `path` when `pathGenerator` changes
  React.useEffect(() => {
    if (divRef.current && features.length) {
      select(divRef.current)
        .select('path')
        .data(features)
        .join('path')
        .attr('d', pathGenerator);
    }
  }, [features, pathGenerator]);

  return (
    <div ref={divRef} data-testid="globe">
      <svg width={width} height={height} fill={landColor}>
        <circle cx={centerX} cy={centerY} r={circleR} fill={oceanColor} />
        <path />
      </svg>
    </div>
  );
}
