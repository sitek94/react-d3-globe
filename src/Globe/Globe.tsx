import * as React from 'react';

import pkg from '../../package.json';
import { createStyles } from '../styles/create-styles';
import { Color } from '../styles/schemes';
import { CountryFeature } from '../types';
import { useCountries } from '../use-countries/use-countries';
import { useProjection } from '../use-projection/use-projection';

let index = 0;
export interface GlobeProps {
  className?: string;
  countryColor?: Color;
  countryHoverColor?: Color;
  oceanColor?: Color;
  size?: number;
}

function Globe(props: GlobeProps) {
  index++;

  const { countryColor, countryHoverColor, oceanColor, size = 400 } = props;

  const width = size;
  const height = size;
  const cx = width / 2;
  const cy = height / 2;
  const r = size / 2;

  const svgRef = React.useRef<SVGSVGElement>(null);
  const { countries } = useCountries();
  const { rotateTo } = useProjection({
    cx,
    cy,
    scale: size / 2,
    svgRef,
    countries,
  });

  const className = `${pkg.name}-${index}`;

  const styles = createStyles(className, {
    countryColor,
    countryHoverColor,
    oceanColor,
  });

  const handleCountryClick = (country: CountryFeature) => {
    rotateTo(country.properties.position);
    console.log(country.properties);
  };

  return (
    <svg ref={svgRef} width={width} height={height} className={className}>
      <style>{styles}</style>
      <circle cx={cx} cy={cy} r={r} className="ocean" />
      {countries.map(country => (
        <g key={country.id}>
          <title>{country.properties.name}</title>
          <path
            className="country"
            onClick={() => handleCountryClick(country)}
          />
        </g>
      ))}
    </svg>
  );
}

export { Globe };
