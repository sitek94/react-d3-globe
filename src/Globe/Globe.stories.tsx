import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Globe, GlobeProps } from './Globe';

// .storybook/preview.js

export default {
  title: 'Globe',
  component: Globe,
  argTypes: {
    size: {
      control: { type: 'range', min: 50, max: 800 },
    },
  },
} as Meta;

const Template: Story<GlobeProps> = args => <Globe {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  size: 400,
};
