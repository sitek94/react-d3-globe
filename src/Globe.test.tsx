import * as React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import Globe from './Globe';

describe('Example', () => {
  it('should match the snapshot', () => {
    const tree = renderer.create(<Globe />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly', () => {
    render(<Globe />);

    expect(screen.getByTestId('globe')).toBeTruthy();
  });
});
