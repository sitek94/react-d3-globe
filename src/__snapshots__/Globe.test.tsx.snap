// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example should match the snapshot 1`] = `
<div
  data-testid="globe"
>
  <svg
    fill="white"
    height={150}
    width={150}
  >
    <circle
      cx={75}
      cy={75}
      r={75}
    />
    <path />
  </svg>
</div>
`;
